### PR TITLE
Refactoring BkTimeable.java

### PR DESCRIPTION
**Pull request by sebing**
Refactored the class according to this article:

http://www.yegor256.com/2015/05/07/ctors-must-be-code-free.htm

Fixing the issue(https://github.com/yegor256/takes/issues/598).

### CODE REVIEW

**Comment by sebing** [comment]
@davvd  Can you find someone to review my changes?

**Comment by davvd** [comment]
@sebing Thanks, let me find someone who can review this pull request

**Comment by davvd** [comment]
@carlosmiranda please review, thanks

**Comment by carlosmiranda** [comment]
@rultor good to merge

File: src/main/java/org/takes/http/BkTimeable.java [hunk 1]
```diff
@@ -130,7 +61,56 @@ private void check() {
...
+        super();
+        this.threads = new ConcurrentHashMap<Thread, Long>(1);
+        this.back = back;
+        this.latency = msec;
+    }
+
+    /**
+     * This method will start the damon thread.
+     */
+    public void startDaemonThread() {
```
Comment (yegor256) [inline src/main/java/org/takes/http/BkTimeable.java]: @sebing oops, that's bad... all public methods must be "Override"... if you have a method that is public and is not implementing any interface - the design is not good. think again.

**Comment by yegor256** [comment]
@sebing something is wrong here, see my comment

File: src/main/java/org/takes/http/BkTimeable.java (hunk 1 above)
Comment (mkordas) [inline src/main/java/org/takes/http/BkTimeable.java]: @yegor256 would you like to make it general rule in Qulice? Or it's just a thing to pay attention to during code reviews?

**Comment by sebing** [comment]
@carlosmiranda can you re-review the code as I fixed the issue reported.

File: src/main/java/org/takes/http/DaemonBack.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * @author Sebin George (sebing@gmail.com)
+ * @version $Id$
+ * @since 1.0
+ */
+public interface DaemonBack extends Back {
+    /**
+     * Start the damon thread.
+     */
+    void startDaemonThread();
```
Comment (carlosmiranda) [inline src/main/java/org/takes/http/DaemonBack.java]: @sebing let's just call this `start`

**Comment by carlosmiranda** [comment]
@sebing one more comment above

**Comment by sebing** [comment]
@carlosmiranda the build timed out. Just checking if it works if i restart.

**Comment by sebing** [comment]
@carlosmiranda the build timed out. Just checking if it works if i restart.

**Comment by sebing** [comment]
@carlosmiranda why both the appveyor and travis is failing, I think they are timed out. Can you help me with this?

**Comment by carlosmiranda** [comment]
@sebing should be fine, let's try to merge

**Comment by carlosmiranda** [comment]
@rultor merge

**Comment by sebing** [comment]
@carlosmiranda hmm.. I would like both the tests in green :) So we know things are good than just my judgement of it is working :)

File: src/main/java/org/takes/http/BkTimeable.java [hunk 3]
```diff
@@ -129,7 +61,54 @@ private void check() {
...
     BkTimeable(final Back back, final long msec) {
-        super(new BkThreads(msec, back));
+        super();
+        this.threads = new ConcurrentHashMap<Thread, Long>(1);
+        this.back = back;
+        this.latency = msec;
     }
 
+    @Override
+    public void start() {
```
Comment (yegor256) [inline src/main/java/org/takes/http/BkTimeable.java]: @sebing I'm curious, why can't we do it all in `run()`? why do we need "start" and "run"?

**Comment by yegor256** [comment]
@sebing I didn't understand the idea behind this "start" method, see above

**Comment by sebing** [comment]
@carlosmiranda the method naming of start in `DaemonBack` is confusing(not just for us)even java :)
So I re-named them. Can you re-review my code?

**Comment by carlosmiranda** [comment]
@sebing I think @yegor256 is asking for an explanation of your solution (he's the project's architect).

**Comment by sebing** [comment]
@carlosmiranda I think, the start method name in Thread and DaemonBack is confusing. @yegor256 's comment made me realize. I think, It is better this way now :)

**Comment by sebing** [comment]
@yegor256 I renamed the method, can you review the code.

**Comment by sebing** [comment]
@yegor256 can we merge this PR?

**Comment by yegor256** [comment]
@sebing it's not about the name of the method :) it's about the design, that is wrong. just introducing a new interface is not a solution. you're doing something that breaks the design of the entire framework. previous solution (which you're trying to change) was in line with the entire design. your code is not. @carlosmiranda don't you see it?

**Comment by sebing** [comment]
@carlosmiranda I need some direction, I am in a catch22 situation :)

BkTimeable should be started as daemon thread, the options I have are all wrong:
1) start the thread in the constructor (which is wrong because `ConstructorOnlyInitializesOrCallOtherConstructors`)
2) Call the public methord implementation, which starts the thread(like the code in FtCLI), but for this we need another interface, we cannot just have public method in BkTimeable.
So I created the interface `DaemonBack` to solve the problem.

Now can you guide me with a third alternative? :)

**Comment by carlosmiranda** [comment]
@sebing @yegor256 sorry guys, I am on vacation right now, didn't have much time to look at the change.

**Comment by carlosmiranda** [comment]
@yegor256 I see it now. @sebing why can't we put the monitor thread start logic in `run()`? It should start the thread, if it's not started yet, like a lazy init. Would be much simpler than what you're trying to do now.

**Comment by sebing** [comment]
@carlosmiranda That makes sense :) I have made the changes can you review them?

**Comment by sebing** [comment]
@yegor256 Can you review the changes?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@ypshenychka please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by ypshenychka** [comment]
@carlosmiranda 
According to our [QA Rules](http://at.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.

Only one issue was found during code review. 
Please confirm that you'll try to find at least three problems while future reviews.

**Comment by carlosmiranda** [comment]
@ypshenychka confirmed

**Comment by ypshenychka** [comment]
@carlosmiranda Thank you.

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thanks a lot, next time everybody should try to make it better

**Comment by davvd** [comment]
@carlosmiranda 10 mins was added to the account of @ypshenychka (for QA review), in transaction 79198945... **24 mins** sent to your balance (ID `79198959`), many thanks! It took 357 hours and 4 mins.... extra minutes for review comments (c=9)... +24 added to your rating, current score is: [+3159](http://www.netbout.com/b/35023?open=rating)

**Pull request merged**